bindgen/src/formatter.ts

```diff
@@ -22,51 +22,49 @@ import cp from "child_process";
 import { extend } from "./debug";
 const debug = extend("format");
 
-const FORMATTERS: Record<string, string[]> = {
-  eslint: ["npx", "eslint", "--fix", "--format=stylish"],
-  "clang-format": ["npx", "clang-format", "-i"],
-  // "typescript-checker": ["npx", "tsc", "--noEmit"],
-};
+export type Formatter = (cwd: string, filePaths: string[]) => void;
 
-export function addFormatter(name: string, commandAndArgs: string[]) {
-  if (name in FORMATTERS) {
-    throw new Error(`Cannot add ${name} as it's already there`);
-  } else {
-    FORMATTERS[name] = commandAndArgs;
+export function executeCommand(cwd: string, command: string, ...args: string[]) {
+  console.log(chalk.dim(command, ...args));
+  const result = cp.spawnSync(command, args, {
+    cwd,
+    encoding: "utf8",
+    stdio: "inherit",
+    shell: true,
+  });
+  if (result.error) throw result.error;
+  if (result.status) {
+    throw new Error(`Exited with status ${result.status}`)
   }
 }
 
-export function getFormatterNames() {
-  return Object.keys(FORMATTERS);
+export function createCommandFormatter(formatterName: string, [command, ...args]: string[]): Formatter {
+  const formatter: Formatter = (cwd, filePaths) => {
+    executeCommand(cwd, command, ...args, ...filePaths);
+  };
+  Object.defineProperty(formatter, "name", { value: formatterName });
+  return formatter;
 }
 
-export type FormatterName = string;
+export const clangFormatter = createCommandFormatter("clang", ["npx", "clang-format", "-i"]);
 
 export class FormatError extends Error {
-  constructor(formatterName: string, filePaths: string[]) {
-    super(`Failure when running the '${formatterName}' formatter on ${JSON.stringify(filePaths)}`);
+  constructor(formatterName: string, filePaths: string[], cause: Error) {
+    super(`Failure when running the '${formatterName}' formatter on ${JSON.stringify(filePaths)}: ${cause.message}`);
```

Comment:
Should we be passing `{cause}` as a second parameter to `super()`? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause#rethrowing_an_error_with_a_cause

Reply:
Oh yeah - I forgot that got added 👍

Reply:
Oh - I have to raise the tsconfig target to `es2022`, but that's probably okay?

Reply:
Ugh, that's annoying since I think the `cause` option is supposed to be backwards compatible and ignored by old JS implementations. But it should be fine.

Ideally TS would let you set a min JS standard, but still let you use backward-compatible features from newer standards, and do the right thing if you use eg `if (globalThis.WeakRef)` to check if you have access to newer features. But that isn't the world we live in...